### ISSUE TITLE
headless Ebitengine

### ISSUE THREAD

**Issue by hajimehoshi** [hajimehoshi/ebiten#353]
Now `(*Image).At` panics if this is called before `ebiten.Run` since OpenGL context is initialized only after `ebiten.Run` (or `ebiten.RunWithoutMainLoop` on mobiles), and this is a restriction for mobiles. If software rendering is implemented and used for such `At` callings, Ebiten would be available without any window system.

**Comment by hajimehoshi** [comment]
Hmm, I forgot the reason why OpenGL is initialized only after `ebiten.RunWithoutMainLoop`

**Comment by hajimehoshi** [comment]
https://github.com/hajimehoshi/ebiten/issues/272#issuecomment-319432307

Re @jakecoffman

Actually, Ebiten works on servers as long as the server as X or other requirements. Ebiten is tested on servers (TravisCI). Is this what you want?

This issue is about the problem that `At` doesn't work without `Run`.

**Comment by jakecoffman** [comment]
I think I misunderstood that ebiten could not run on a headless server but it sounds like it can, so no problems!

**Comment by silbinarywolf** [comment]
Related to this somewhat. Is there a way to build/run Ebiten without GL/GLFW?
The reasons I'd want this are:
- Cross compilation from Windows won't give errors. (CGO + Cross compiling on Windows doesn't work well)
- Server code will theoretically do less work as the draw calls / etc would be stubbed / emptied.
- GLFW won't ever fail to open the display / server doesn't need any windowing system installed or running, so I assume that would mean less CPU overhead + lower server costs.

I'm trying to run my experimental game server stuff in a Docker container but I get the following crash message.

```
app_1  | 2018/06/09 06:55:49 PlatformError: X11: Failed to open display :99.0
app_1  | panic: NotInitialized: The GLFW library is not initialized
app_1  |
app_1  | goroutine 1 [running, locked to thread]:
app_1  | github.com/go-gl/glfw/v3.2/glfw.acceptError(0x0, 0x0, 0x0, 0x1, 0xc42005de18)
app_1  |        /go/src/github.com/go-gl/glfw/v3.2/glfw/error.go:173 +0x1d4
app_1  | github.com/go-gl/glfw/v3.2/glfw.panicError()
app_1  |        /go/src/github.com/go-gl/glfw/v3.2/glfw/error.go:184 +0x32
app_1  | github.com/go-gl/glfw/v3.2/glfw.WindowHint(0x20004, 0x0)
app_1  |        /go/src/github.com/go-gl/glfw/v3.2/glfw/window.go:229 +0x38
app_1  | github.com/hajimehoshi/ebiten/internal/ui.initialize(0x95fc60, 0xc42004e350)
app_1  |        /go/src/github.com/hajimehoshi/ebiten/internal/ui/ui_glfw.go:83 +0x4e
app_1  | github.com/hajimehoshi/ebiten/internal/ui.init.0()
app_1  |        /go/src/github.com/hajimehoshi/ebiten/internal/ui/ui_glfw.go:74 +0x27
game_app_1 exited with code 2
```

Dockerfile:
```
FROM golang:1.10.2

WORKDIR /go/src/github.com/silbinarywolf/game
COPY . .

RUN apt-get update && \
	apt-get install -y libx11-dev \
					libgl1-mesa-dev \
					libasound2-dev \
					libgles2-mesa-dev \
					libalut-dev \
					libxcursor-dev \
					libxi-dev \
					libxinerama-dev \
					libxrandr-dev \
					libxxf86vm-dev

RUN go get -d -v ./...
RUN go install -v ./...

# This code isn't actually on GitHub... it's private.
CMD ["github.com/silbinarywolf/game"]

RUN go build
```

**Comment by silbinarywolf** [comment]
I had a look at these fews lines:

- export DISPLAY=:99.0
- sh -e /etc/init.d/xvfb start

When I try to run the latter, I get a "not found" error. So I'm assuming I'd need to install some sort of X Windows package into the Docker container for this.

I think I'll explore tidying up my game engine code and implementing it at my code level. I just wanted to make sure there wasn't a relatively straight forward way of getting this to work with Ebiten out of the box!

**Comment by hajimehoshi** [comment]
> I just wanted to make sure there wasn't a relatively straight forward way of getting this to work with Ebiten out of the box!

Yeah, the less prerequisites, the happier everyone would be :-) For the current architecture with GLFW, I guess it is impossible to build games without X, unfortunately.

**Comment by hajimehoshi** [comment]
My impression:

* If just hiding the window is enough, this can be achieved very easily.
* If we need to remove the dependency on X, it would be much harder.

**Comment by kyeett** [comment]
Hi! The dependency is not a problem for me. It would be nice to be able to turn these things off
1. Window popping up and taking over focus (highest prio)
2. Window popping up at all.

**Comment by hajimehoshi** [comment]
@kyeett Would #769 (Launching without focusing) help?

**Comment by hajimehoshi** [comment]
The purpose seems the same as #790

**Comment by hajimehoshi** [comment]
Headless Ebitengine is a little broater notion than software rendering (#790), so let me reopen this

**Comment by hajimehoshi** [comment]
On second thought, I'll create another issue.

**Comment by hajimehoshi** [comment]
With the context of the vm (#3438), a headless mode is different from #790
